marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs

```diff
@@ -287,11 +303,19 @@ buildApplyInputsConstraintsV1 systemStart eraHistory marloweOutput invalidBefore
     requiredParties = Set.fromList $ for redeemer $ marloweInputParty >>> maybeToList
     roleAssetId = toAssetId currencySymbol
 
-  invalidBefore' <- lift $ utcTimeToSlotNo invalidBefore
+    tipSlot' = toCardanoSlotNo tipSlot
+
+  invalidBefore' <- lift $ maybe (pure tipSlot') utcTimeToSlotNo invalidBefore
```

Comment:
We'll need extensive testing of Runtime's time-to-slot conversion, to make sure it is consistent with the validator's https://github.com/input-output-hk/marlowe-cardano/blob/92916702040c995ac51ee541fb6c28a382102c65/marlowe/src/Language/Marlowe/Scripts.hs#L159-L168.

We've had problems in the past where a small roundoff in the time interval propagates into the `minTime` in the datum differently in the on- and off-chain code, with the result that the transaction fails due to the on- and off-chain outputs having different `minTime`.

The code here looks correct, but it's hard enough to reason about this that we'll need test cases. I'll create a Jira ticket.